### PR TITLE
Merge `.pyi` type stubs inline

### LINKED ISSUE #536: [PROPOSAL] Merge .pyi types inline
### What are you proposing?

We have the codebase littered with .pyi files. Is it time to merge types into Python code?

### What problems are you trying to solve?

Updating types in multiple places is error prone.

### What is the developer experience going to be?

No more .pyi files.

### PR DESCRIPTION
**Pull request by dblock**
### Description

Merge .pyi files inline for those files that were previously covered by mypy. Many `Any` that should not be, but it gets us started to have a fully typed library that is edited at the same time as code. This should greatly improve code clarity, maintainability, and enable better static analysis.

I wasn't sure whether this is a good idea, so I found some other projects merging stubs.

- https://github.com/pytorch/pytorch/issues/102428 says that stubs mask the existence of methods, which can lead to spurious type errors in child classes of these classes, and turns off mypy for functions whose type hints are defined only in the stubs, since mypy doesn't run on functions/methods without inline type hints.
- https://github.com/django/deps/pull/65

### Issues Resolved

Closes #536.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dblock** [comment]
@saimedhi this is ready for review, FYI

**Review by saimedhi — COMMENTED** [review]
@dblock, I've reviewed the PR, and from what I understand, API types are added automatically through the generator, while for tests and other files, we add them manually. I've made a few requested changes, and if you agree with them, please make the changes. Thank you for your contribution!

File: benchmarks/bench_async.py [hunk 1]
```diff
@@ -71,34 +71,34 @@ async def test_async(client_count=1, item_count=1):
     await asyncio.gather(*[client.close() for client in clients])
 
 
-def test(item_count=1, client_count=1):
+def test(item_count: int = 1, client_count: int = 1) -> None:
```
[the remark this answers is not in the document]
Comment (dblock) [reply benchmarks/bench_async.py:74]: There's another 1000 errors from the missing files that were not previously checked by mypy, including benchmarks and tests. Could we merge this one and I'll work on the rest?

File: benchmarks/bench_async.py [hunk 2]
```diff
@@ -22,7 +22,7 @@
 item_count = 100
 
 
-async def index_records(client, item_count):
+async def index_records(client, item_count) -> None:
```
[the remark this answers is not in the document]
Comment (dblock) [reply benchmarks/bench_async.py:25]: These weren't covered by mypy before.

```
-    session.run("mypy", "--strict", "opensearchpy/")
+    session.run("mypy", "--strict", *SOURCE_FILES)
```

Adds another 1000 errors 😢 I can do this as a followup.

**Comment by dblock** [comment]
@saimedhi Thanks for the CR! I edited the description to say "Merge .pyi files inline for those files that were previously covered by mypy." Opened https://github.com/opensearch-project/opensearch-py/issues/564 to add types to the rest of the code.

File: benchmarks/bench_async.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dblock) [reply benchmarks/bench_async.py:25]: https://github.com/opensearch-project/opensearch-py/issues/564

**Review by saimedhi — APPROVED** [review]

**Pull request merged**